libopenage/rng/rng.cpp

```diff
@@ -48,58 +49,93 @@ void RNG::seed(uint64_t val) {
 
 
 /**
- * murmurhash3 bit avalancher - output is uniformly distributed
- * and changing one bit of input results in ~1/2 of output bits changed
- * This function is a good way of mixing up an input,
- * or mixing together two results with hash64(val1 ^ val2)
- *
- * https://en.wikipedia.org/wiki/MurmurHash
- */
-static uint64_t hash64(uint64_t val) {
-	// The two magic numbers have been determined heuristically.
-	// They avalanche all bits in val to within 0.25% bias.
-
-	val ^= val >> 33;
-	val *= 0xff51afd7ed558ccd;
-	val ^= val >> 33;
-	val *= 0xc4ceb9fe1a85ec53;
-	return val ^ (val >> 33);
+* Rotate left no carry
+*
+* http://blog.regehr.org/archives/1063
+*/
+inline uint64_t rotl64(uint64_t x, uint64_t n)
+{
+	assert (n<64);
+	return (x<<n) | (x>>(-n&63));
 }
 
-
-/*
- * reinterpret_casts 8 chars from a buffer as a uint64_t.
- */
-static inline uint64_t from_cptr(const char *data) {
-	uint64_t result;
-	memcpy(reinterpret_cast<char *>(&result), data, sizeof(result));
-	return result;
+// Maybe this should be a lambda inside of siphash()?
+inline void sipround(uint64_t *v) {
+	v[0] += v[1];
+	v[1] = rotl64(v[1], 13);
+	v[1] ^= v[0];
+	v[0] = rotl64(v[0], 32);
+	v[2] += v[3];
+	v[3] = rotl64(v[3], 16);
+	v[3] ^= v[2];
+	v[0] += v[3];
+	v[3] = rotl64(v[3], 21);
+	v[3] ^= v[0];
+	v[2] += v[1];
+	v[1] = rotl64(v[1], 17);
+	v[1] ^= v[2];
+	v[2] = rotl64(v[2], 32);
 }
 
 
-/*
- * hashes an initial seed into the base state
- */
-static void init_hash(uint64_t *state, uint64_t seed) {
-	// Mix the initial seed into data for each state value
-	state[0] = hash64(seed);
-	state[1] = hash64(seed ^ state[0]);
-}
+// I don't know where to put this, or what value should we use
+static const uint64_t siphash_key[2]{0x0706050403020100ull, 0x0f0e0d0c0b0a0908ull};
```

Comment:
As I understood it, siphash can (and should) be used with a secret key. This means we should not hardcode it, instead I'd say we wrap siphash in another class, which is initialized by this key.
The key is then determined either fixed (e.g. for consistent map generation) or random (e.g. for each game).